File: graph/Number_of_provinces.cpp
/*Author : Raj Roy*/

#include<bits/stdc++.h>
using namespace std;

void dfs(int current,vector<int> adj[],int vis[]){
	vis[current] = 1;
	for(auto it : adj[current]){
		if(!vis[it]){
			dfs(it,adj,vis);
		}
	}
}

int main(){
	int n , m;
	cin >> n >> m;
	vector<int> adj[n + 1];
	int vis[n+1];
	for(int i = 0; i < m; i++){
		int u , v;
		cin >> u >> v;
		adj[u].push_back(v);
		adj[v].push_back(u);
	}
	
	int count = 0;
	for(int i = 0; i < n; i++){
		if(!vis[i]){
			count++;
			dfs(i,adj,vis);
			
		}
	}
	cout << count << endl;
}

File: GRAPH/Number_of_provinces.java
package GRAPH;
import java.util.*;
public class Number_of_provinces {
    public static void main(String[] args) {
        //Take input in the adjacency matrix;
        //return findCircleNum method to get the answer;
    }
    public int findCircleNum(int[][] isConnected) {

        //Adjacency List;
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for(int i = 0; i < isConnected.length; i++){
            adj.add(new ArrayList<>());
        }

        //Method to convert the adjacency matrix to adjacency List;
        for(int i = 0; i < isConnected.length; i++){
            for(int j = 0; j < isConnected[i].length; j++){
                if(isConnected[i][j] == 1 && i != j){
                    adj.get(i).add(j);
                    adj.get(j).add(i);
                }
            }
        }

        int vis[] = new int[isConnected.length + 1];
        Arrays.fill(vis,0);
        int count = 0;
        for(int i = 0;i  < isConnected.length; i++){
            if(vis[i] == 0){
                count++;
                dfs(i,adj,vis);

            }
        }
        return count;
    }

    //dsf function;
    public static void dfs(int current, ArrayList<ArrayList<Integer>> adj,int vis[]){
        vis[current] = 1;
        for(int ele : adj.get(current)){
            if(vis[ele] == 0 ){
                dfs(ele,adj,vis);
            }
        }
    }
}
